### PR TITLE
Collab Dialogues (Add a "block ownership" mode to the collab editor)

### PR DESCRIPTION
**Pull request by jimrandomh**
This is a prototype for a way of running debates. It is not production-ready.

This makes a mode for CkEditor5 collaborative-editing documents where users can have ownership over paragraphs, which prevents other users from editing those paragraphs. You own a paragraph if (1) you are the post author or are explicitly shared (not link-shared), and (2) the paragraph starts with "Yourname:". If someone tries to edit your paragraph, it will but them in Suggesting mode, and they can't approve their own suggestions.

A paragraph can be edited by anyone if it doesn't start with a name and a ':', or it starts with the name of someone who isn't explicitly shared on the post.

This mode is only active if you put "&blockOwnership=true" in the URL of the post-edit page, eg

  http://localhost:3000/editPost?postId=POSTIDHERE&blockOwnership=true.

Each participant must do this manually. If a user doesn't have block-ownership mode turned on or is using a different branch to edit the post, and they have write access to the post, then they can edit every block inside the post even ones that are owned by others.

This also adds a button to the bottom of the post editor, labelled "Add Message". This inserts a paragraph at the end of the document containing "Username:".

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1205551925773701) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by Discordius** [comment]
Note: We used this PR as the branching off-point for the whole new dialogue feature and architecture.

File: packages/lesswrong/components/editor/EditLinkpostUrl.tsx [hunk 1]
```diff
@@ -67,7 +67,7 @@ const EditLinkpostUrl = ({
       <Input
         inputRef={inputRef}
         className={classes.input}
-        value={(document && document[path]) || defaultValue || ""}
+        value={(document[path]) || defaultValue || ""}
```
Comment (b0b3rt) [inline packages/lesswrong/components/editor/EditLinkpostUrl.tsx:70]: Is this change safe?

File: packages/lesswrong/components/editor/EditTitle.tsx [hunk 2]
```diff
@@ -70,7 +64,7 @@ const EditTitle = ({document, value, path, placeholder, updateCurrentValues, cla
   }, [document, updatePost, lastSavedTitle, flash])
 
   return <Input
-    className={classNames(classes.root, {[classes.question]: question && !isEAForum})}
+    className={classes.root}
```
Comment (b0b3rt) [inline packages/lesswrong/components/editor/EditTitle.tsx:67]: We're removing question-specific styling?

File: packages/lesswrong/components/dropdowns/posts/PostActions.tsx [hunk 3]
```diff
@@ -77,10 +77,10 @@ const PostActions = ({post, closeMenu, includeBookmark=true, classes}: {
       />
       <NotifyMeDropdownItem
         document={post}
-        enabled={!!post.debate}
+        enabled={!!post.collabEditorDialogue}
```
Comment (b0b3rt) [inline packages/lesswrong/components/dropdowns/posts/PostActions.tsx:80]: Note that this means users can no longer subscribe to old dialogues

File: packages/lesswrong/components/editor/EditorFormComponent.tsx [hunk 4]
```diff
@@ -284,7 +284,7 @@ export const EditorFormComponent = ({form, formType, formProps, document, name,
       });
       const cleanupSuccessForm = context.addToSuccessForm((result: any, form: any, submitOptions: any) => {
         getLocalStorageHandlers(currentEditorType).reset();
-        if (editorRef.current && !submitOptions?.redirectToEditor) {
+        if (editorRef.current && !submitOptions?.redirectToEditor && !isCollabEditor) {
```
Comment (b0b3rt) [inline packages/lesswrong/components/editor/EditorFormComponent.tsx:287]: What is this change intended to do/prevent/etc?

File: packages/lesswrong/components/posts/NewDialogueDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+      flash("You must choose a title");
+      return;
+    }
+    const createResult = await createPost({
+      data: {
+        title,
+        draft: true,
+        collabEditorDialogue: true,
+        coauthorStatuses: participants.map(userId => ({userId, confirmed: false, requested: false})),
+        shareWithUsers: participants,
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/NewDialogueDialog.tsx:67]: This is safe because `createDialogue` gets recreated (on rerender) each time `participants` are updated via `setParticipants`?

File: packages/lesswrong/components/posts/PostSubmit.tsx [hunk 6]
```diff
@@ -92,7 +92,7 @@ const PostSubmit = ({
   const requireConfirmation = isLW && collectionName === 'Posts' && !!document.debate;
 
   const onSubmitClick = requireConfirmation ? submitWithConfirmation : submitWithoutConfirmation;
-
+
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostSubmit.tsx:95]: note: same, please just take the time to not commit changes like this

File: packages/lesswrong/components/posts/PostsEditForm.tsx [hunk 7]
```diff
@@ -65,6 +63,7 @@ const PostsEditForm = ({ documentId, classes }: {
   });
   const rateLimitNextAbleToPost = userWithRateLimit?.rateLimitNextAbleToPost
     
+
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsEditForm.tsx:66]: unnecessary diff

File: packages/lesswrong/components/dropdowns/posts/PostActions.tsx (hunk 3 above)
Comment (Discordius) [reply packages/lesswrong/components/dropdowns/posts/PostActions.tsx:80]: Yeah, that seems fine

File: packages/lesswrong/components/posts/PostsEditForm.tsx [hunk 8]
```diff
@@ -111,6 +110,11 @@ const PostsEditForm = ({ documentId, classes }: {
     </div>
   }
   
+  let removedFields = []
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsEditForm.tsx:113]: this should be `const`

File: packages/lesswrong/components/editor/EditTitle.tsx (hunk 2 above)
Comment (Raemon) [reply packages/lesswrong/components/editor/EditTitle.tsx:67]: EA Forum removed it, and I think the question-specific-styling only really made sense when there was a different Question Modal with an overall smaller UI. We changed awhile ago from nudging people towards shorter questions towards lager ones.

Comment (Raemon) [reply packages/lesswrong/components/editor/EditTitle.tsx:67]: (This was part of an overall sub-process of "bring our Post Page styling in line with the EA Forum tab-bar, since it made the page simpler and gave us more wiggle room for adding Dialogue specific elements)

File: packages/lesswrong/components/editor/EditorFormComponent.tsx (hunk 4 above)
Comment (Discordius) [reply packages/lesswrong/components/editor/EditorFormComponent.tsx:287]: If you don't make this change, then we try to call `setData` during a collaborative editing session, which crashes the page.

File: packages/lesswrong/components/posts/dialogues/DialogueEditorUI.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+});
+
+export const DialogueEditorUI = ({classes}: {
+  classes: ClassesType,
+}) => {
+  return <div className={classes.root}>
+    <div>Dialogue Editor</div>
+    <ul className={classes.info}>
+      <li>You can edit your responses afterwards.</li>
+      <li>You can see each other's responses as you type them</li>
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/dialogues/DialogueEditorUI.tsx:30]: missing period at the end of the sentence

File: packages/lesswrong/components/posts/dialogues/DialogueSubmit.tsx [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  },
+  cancelButton: {
+  },
+  draft: {
+  },
+  feedback: {
+  }
+});
+
+const isEAForum = forumTypeSetting.get() === "EAForum"
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/dialogues/DialogueSubmit.tsx:55]: There's a global we should import instead of redefining it here

File: packages/lesswrong/components/posts/NewDialogueDialog.tsx (hunk 5 above)
Comment (Discordius) [reply packages/lesswrong/components/posts/NewDialogueDialog.tsx:67]: Looks like that to me. Seems fine, I think?

File: packages/lesswrong/components/posts/dialogues/DialogueSubmit.tsx [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+DialogueSubmit.contextTypes = {
+  updateCurrentValues: PropTypes.func,
+  addToSuccessForm: PropTypes.func,
+  addToSubmitForm: PropTypes.func,
+  submitForm: PropTypes.func
+}
+
+
+// HACK: Cast DialogueSubmit to hide the legacy context arguments, to make the type checking work
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/dialogues/DialogueSubmit.tsx:146]: Hm, was suspicious, but looks like this was necessary in `PostSubmit` as well.

File: packages/lesswrong/components/editor/EditLinkpostUrl.tsx (hunk 1 above)
Comment (Raemon) [reply packages/lesswrong/components/editor/EditLinkpostUrl.tsx:70]: I removed it because I was getting a "document is null-checked here, but not earlier" warning. But, I'm not actually that confident it should resolve in this direction.

Comment (Raemon) [reply packages/lesswrong/components/editor/EditLinkpostUrl.tsx:70]: See line 50

File: packages/lesswrong/components/users/ViewSubscriptionsPage.tsx [hunk 12]
```diff
@@ -142,19 +142,37 @@ const ViewSubscriptionsPage = ({classes}: {
       renderDocument={(post: PostsList) => post.title}
       noSubscriptionsMessage="You are not subscribed to comments on any posts."
     />
-
+
```
Comment (b0b3rt) [inline packages/lesswrong/components/users/ViewSubscriptionsPage.tsx:145]: how do these happen????

File: packages/lesswrong/components/users/ViewSubscriptionsPage.tsx [hunk 13]
```diff
@@ -142,19 +142,37 @@ const ViewSubscriptionsPage = ({classes}: {
...
       title="Subscribed to Dialogues (as a reader)"
       collectionName="Posts"
-      subscriptionType="newDebateComments"
+      subscriptionType="newPublishedDialogueMessages"
       fragmentName="PostsList"
       renderDocument={(post: PostsList) => post.title}
       noSubscriptionsMessage="You are not subscribed to any dialogues as a reader."
     />
-
+
```
Comment (b0b3rt) [inline packages/lesswrong/components/users/ViewSubscriptionsPage.tsx:154]: fr

File: packages/lesswrong/lib/collections/posts/schema.tsx [hunk 14]
```diff
@@ -1361,13 +1361,14 @@ const schema: SchemaType<DbPost> = {
...
     control: "CoauthorsListEditor",
-    group: formGroups.coauthors
+    group: formGroups.coauthors,
   },
   'coauthorStatuses.$': {
     type: new SimpleSchema({
       userId: String,
       confirmed: Boolean,
       requested: Boolean,
+
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/posts/schema.tsx:1371]: ???

File: packages/lesswrong/lib/collections/posts/schema.tsx [hunk 15]
```diff
@@ -2367,12 +2374,14 @@ const schema: SchemaType<DbPost> = {
...
   
+  // This is basically deprecated. We now have them enabled by default
+  // for all users. Leaving this field for legacy reasons.
   sideCommentVisibility: {
     type: String,
     optional: true,
     control: "select",
     group: formGroups.advancedOptions,
-    hidden: (props) => props.eventForm || !userHasSideComments(props.currentUser),
+    hidden: true,
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/posts/schema.tsx:2383]: Was this diff rolled in deliberately?

File: packages/lesswrong/lib/collections/posts/views.ts [hunk 16]
```diff
@@ -167,7 +167,7 @@ export const sortings: Record<PostSortingMode,MongoSelector<DbPost>> = {
  * as it is *inclusive*. The parameters callback that handles it outputs
  * ~ $lt: before.endOf('day').
  */
-Posts.addDefaultView((terms: PostsViewTerms, _, context: ResolverContext) => {
+Posts.addDefaultView((terms: PostsViewTerms, _, context?: ResolverContext) => {
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/posts/views.ts:170]: Did we discover that context is optional here?

File: packages/lesswrong/components/posts/dialogues/DialogueSubmit.tsx (hunk 10 above)
Comment (Raemon) [reply packages/lesswrong/components/posts/dialogues/DialogueSubmit.tsx]: Fixed

File: packages/lesswrong/lib/collections/posts/schema.tsx (hunk 15 above)
Comment (Raemon) [reply packages/lesswrong/lib/collections/posts/schema.tsx:2383]: We deliberately decided to remove the option to change side-comment display. But, occurs to me I should inform EA Forum about it.

File: packages/lesswrong/lib/collections/posts/views.ts (hunk 16 above)
Comment (Raemon) [reply packages/lesswrong/lib/collections/posts/views.ts:170]: I just tried removing the diff and didn't see any errors, but it was added by @jimrandomh I suspect he had some kind of reason.

File: packages/lesswrong/lib/collections/posts/helpers.ts [hunk 17]
```diff
@@ -10,7 +10,7 @@ import moment from '../../moment-timezone';
 import { max } from "underscore";
 import { TupleSet, UnionOf } from '../../utils/typeGuardUtils';
 
-export const postCategories = new TupleSet(['post', 'linkpost', 'question'] as const);
+export const postCategories = new TupleSet(['post', 'linkpost', 'question', 'dialogue'] as const);
```
Comment (Raemon) [inline packages/lesswrong/lib/collections/posts/helpers.ts:13]: I thought this change was useful, but I think upon reflection it wasn't. Planning to revert.

File: packages/lesswrong/server/resolvers/postResolvers.ts [hunk 18]
```diff
@@ -77,15 +77,15 @@ augmentFieldsDict(Posts, {
...
+        return getDialogueResponseIds(post).length
       }
     }
   },
   unreadDebateResponseCount: {
     resolveAs: {
       type: 'Int',
       resolver: async (post, _, context) => {
-        if (!post.debate) return 0;
+        if (!post.collabEditorDialogue) return 0;
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/postResolvers.ts:88]: This is a backwards-breaking change for existing dialogues.  If we're reusing the same resolver field for both of them, then we need to preserve previous logic for old dialogues.

**Review by b0b3rt — APPROVED** [review]
